src/nrnmpi/nrnmpi_dynam.c

```diff
@@ -73,44 +77,103 @@ static void* load_nrnmpi(const char* name, char* mes) {
 #if DARWIN
-	sprintf(pmes, "Try loading openmpi\n");
-	void* handle = load_mpi("libmpi.dylib", pmes+strlen(pmes));
+	sprintf(pmes, "Try loading libmpi\n");
+	handle = load_mpi("libmpi.dylib", pmes+strlen(pmes));
+#if defined(NRNCMAKE)
 	if (handle) {
+		/* loaded but is it openmpi or mpich */
+		if (dlsym(handle, "ompi_mpi_init")) { /* it is openmpi */
 		/* see man dyld */
+			if (!load_nrnmpi("@loader_path/libnrnmpi_ompi.dylib", pmes+strlen(pmes))) {
+				return pmes;
+			}
+		}else{ /* must be mpich. Could check for MPID_nem_mpich_init...*/
+			if (!load_nrnmpi("@loader_path/libnrnmpi_mpich.dylib", pmes+strlen(pmes))) {
+				return pmes;
+			}
+		}
+	}else{
+		ismes = 1;
+sprintf(pmes+strlen(pmes), "Is openmpi or mpich installed? If not in default location, need a LD_LIBRARY_PATH.\n");
+	}
+#else /* autotools has only a libnrnmpi that is likely only for openmpi */
+	if (handle) {
 		if (!load_nrnmpi("@loader_path/libnrnmpi.dylib", pmes+strlen(pmes))) {
 			return pmes;
 		}
 	}else{
 		ismes = 1;
 sprintf(pmes+strlen(pmes), "Is openmpi installed? If not in default location, need a LD_LIBRARY_PATH.\n");
 	}
+#endif /* not defined (NRNCMAKE) */
 #else /*not DARWIN*/
 #if defined(MINGW)
 	sprintf(pmes, "Try loading msmpi\n");
-	void* handle = load_mpi("msmpi.dll", pmes+strlen(pmes));
+	handle = load_mpi("msmpi.dll", pmes+strlen(pmes));
 	if (handle) {
+#if defined(NRNCMAKE)
+		if (!load_nrnmpi("libnrnmpi_msmpi.dll", pmes+strlen(pmes))){
+#else
 		if (!load_nrnmpi("libnrnmpi.dll", pmes+strlen(pmes))){
+#endif
 			return pmes;
 		}
 	}else{
 		ismes = 1;
 		return pmes;
 	}
-#else /*not MINGW*/
-	sprintf(pmes, "Try loading openmpi\n");
-	void* handle = load_mpi("libmpi.so", pmes+strlen(pmes));
-	if (handle) {
+#else /*not MINGW so must be __linux__*/
+
+	/**
+	 * libmpi.so is not standard but used by most of the implemenntation
+	 * (mpich, openmpi, intel-mpi, parastation-mpi, hpe-mpt) but not cray-mpich.
+	 * we first load libmpi and then libmpich.so as a fallaback for cray system.
+	 */
+	sprintf(pmes, "Try loading libmpi\n");
+	handle = load_mpi("libmpi.so", pmes+strlen(pmes));
+	if (!handle) {
+	    sprintf(pmes, "Try loading libmpi and libmpich\n");
+	    handle = load_mpi("libmpich.so", pmes+strlen(pmes));
+	}
+
 #if defined(NRNCMAKE)
-/* with CMAKE the problem of Python launch on LINUX not resolving
-   variables from already loaded shared libraries has returned.
-*/
-if (!dlopen("libnrniv.so", RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL)) {
-	fprintf(stderr, "Did not promote libnrniv.so to RTLD_GLOBAL: %s\n", dlerror());
-}
-#endif
+	if (handle) {
+		/* with CMAKE the problem of Python launch on LINUX not resolving
+		   variables from already loaded shared libraries has returned.
+		*/
+		if (!dlopen("libnrniv.so", RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL)) {
+			fprintf(stderr, "Did not promote libnrniv.so to RTLD_GLOBAL: %s\n", dlerror());
+		}
+
+		/* safest to use full path for libnrnmpi... */
+		const char* prefix = path_prefix_to_libnrniv();
```

Comment:
is the returned `prefix` is freed?

Reply:
no. path_prefix_to_libnrniv keeps a singleton char* to return immediately if called again.